experimental/learn_descriptors/hloc_helper.py

```diff
@@ -0,0 +1,254 @@
+from pathlib import Path
+from dataclasses import dataclass
+import pycolmap
+
+import matplotlib.pyplot as plt
+from hloc import (
+    extract_features,
+    match_features,
+    reconstruction,
+    visualization,
+    pairs_from_retrieval,
+    pairs_from_exhaustive,
+    localize_sfm,
+)
+from hloc.utils import viz_3d
+from hloc.utils.parsers import parse_image_list
+from typing import Optional, Union, List
+import random
+import pickle
+import numpy as np
+import plotly.graph_objects as go
+
+
+@dataclass
+class HlocHelperConfig:
+    dir_images: Path
+    dir_outputs: Path
+    path_sfm_pairs: Path
+    dir_sfm: Path
+    local_feature_config: dict  # local feature conf
+    matcher_local_config: dict
+    global_feature_config: (
+        dict | None
+    )  # used for reconstruction if exhaustive_match is false and/or localization
+    exhaustive_match: bool = False
+    retrieval_num_matched: int = (
+        30  # number of pairs to generate per image using global features (set by global_feature_config)
+    )
+    path_local_features: Path | None = None
+    path_global_features: Path | None = None
+    path_local_matches_sfm: Path | None = None
+
+    def __post_init__(self):
+        assert isinstance(self.dir_images, Path)
+        assert self.dir_images.exists()
+        assert isinstance(self.dir_outputs, Path)
+        assert isinstance(self.path_sfm_pairs, Path)
+        assert isinstance(self.dir_sfm, Path)
+        assert isinstance(self.local_feature_config, dict)
+        assert isinstance(self.matcher_local_config, dict)
+        assert isinstance(self.exhaustive_match, bool)
+        assert self.exhaustive_match or isinstance(self.global_feature_config, dict)
+        assert self.exhaustive_match or isinstance(self.retrieval_num_matched, int)
+        if self.path_local_features is None:
+            self.path_local_features = self.dir_outputs / "features.h5"
+        if self.path_global_features is None:
+            self.path_global_features = self.dir_outputs / "global_features.h5"
+        if self.path_local_matches_sfm is None:
+            self.path_local_matches_sfm = self.dir_outputs / "matches.h5"
+
+
+class HlocHelper:
+    def __init__(
+        self,
+        config: HlocHelperConfig | None = None,
```

Comment:
Is having `config=None` ever valid? If there is some way to refactor this so you always need a valid config, then it might save you some cognitive load.